File: frontend/test-api-connection.mjs
/**
 * Frontend API Connection Test
 * Tests connectivity from frontend to backend API
 * Run with: node test-api-connection.mjs
 */

const BACKEND_URL = process.env.VITE_BACKEND_API_URL || "http://192.168.0.71:8001";

console.log("\n" + "=".repeat(60));
console.log("FRONTEND-BACKEND CONNECTION TEST");
console.log("=".repeat(60));
console.log(`Backend URL: ${BACKEND_URL}\n`);

async function testEndpoint(method, endpoint) {
  try {
    const url = new URL(`${BACKEND_URL}/${endpoint.replace(/^\//, "")}`);
    console.log(`Testing ${method} ${endpoint}...`);

    const response = await fetch(url.toString(), {
      method: method,
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (!response.ok) {
      console.log(`  ✗ Status ${response.status}`);
      return false;
    }

    const data = await response.json();
    if (Array.isArray(data)) {
      console.log(`  ✓ Status 200 OK (returned ${data.length} items)`);
    } else {
      console.log(`  ✓ Status 200 OK`);
      console.log(`  Response:`, data);
    }
    return true;
  } catch (error) {
    console.log(`  ✗ Error: ${error.message}`);
    return false;
  }
}

async function testBackendAvailability() {
  console.log("\n" + "=".repeat(60));
  console.log("BACKEND AVAILABILITY TEST");
  console.log("=".repeat(60));

  try {
    console.log(`Connecting to ${BACKEND_URL}...`);
    const response = await fetch(`${BACKEND_URL}/`, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (response.ok) {
      const data = await response.json();
      console.log("✓ Backend is running!");
      console.log(`✓ Response: `, data);
      return true;
    } else {
      console.log(`✗ Backend returned status ${response.status}`);
      return false;
    }
  } catch (error) {
    console.log(`✗ Cannot connect to backend at ${BACKEND_URL}`);
    console.log(`   Error: ${error.message}`);
    console.log("   Make sure the backend is running:");
    console.log("   cd APMS && python -m uvicorn backend.main:app --reload");
    return false;
  }
}

async function testCORSHeaders() {
  console.log("\n" + "=".repeat(60));
  console.log("CORS HEADERS TEST");
  console.log("=".repeat(60));

  try {
    const response = await fetch(`${BACKEND_URL}/`, {
      method: "GET",
    });

    const corsHeaders = {
      "access-control-allow-origin": response.headers.get("access-control-allow-origin") || "NOT SET",
      "access-control-allow-methods": response.headers.get("access-control-allow-methods") || "NOT SET",
      "access-control-allow-headers": response.headers.get("access-control-allow-headers") || "NOT SET",
    };

    console.log("\nCORS Headers:");
    for (const [header, value] of Object.entries(corsHeaders)) {
      console.log(`  ${header}: ${value}`);
    }

    if (corsHeaders["access-control-allow-origin"] !== "NOT SET") {
      console.log("\n✓ CORS is configured");
      return true;
    } else {
      console.log("\n⚠ CORS headers not found (may be configured for preflight only)");
      return true;
    }
  } catch (error) {
    console.log(`✗ Error checking CORS: ${error.message}`);
    return false;
  }
}

async function testAPIEndpoints(backendAvailable) {
  if (!backendAvailable) {
    console.log("\n(Skipping endpoint tests - backend not available)");
    return {};
  }

  console.log("\n" + "=".repeat(60));
  console.log("API ENDPOINTS TEST");
  console.log("=".repeat(60) + "\n");

  const endpoints = [
    { method: "GET", path: "/tools/" },
    { method: "GET", path: "/projects/" },
    { method: "GET", path: "/interventions/" },
    { method: "GET", path: "/tool-types/" },
  ];

  const results = {};
  for (const { method, path } of endpoints) {
    results[path] = await testEndpoint(method, path);
  }

  return results;
}

async function main() {
  const backendAvailable = await testBackendAvailability();
  await testCORSHeaders();
  const endpointResults = await testAPIEndpoints(backendAvailable);

  // Summary
  console.log("\n" + "=".repeat(60));
  console.log("TEST SUMMARY");
  console.log("=".repeat(60));

  const results = {
    "Backend Running": backendAvailable,
    "Endpoints Accessible": Object.values(endpointResults).every(v => v === true),
  };

  for (const [test, passed] of Object.entries(results)) {
    const status = passed ? "✓ PASS" : "✗ FAIL";
    console.log(`${status}: ${test}`);
  }

  const allPassed = Object.values(results).every(v => v === true);

  console.log("\n" + "=".repeat(60));
  if (allPassed) {
    console.log("✓ ALL TESTS PASSED!");
    console.log("Frontend can connect to backend successfully!");
  } else {
    console.log("✗ SOME TESTS FAILED");
    console.log("\nTroubleshooting:");
    if (!backendAvailable) {
      console.log("  - Start the backend server:");
      console.log("    cd APMS && python -m uvicorn backend.main:app --reload");
      console.log("  - Or set VITE_BACKEND_API_URL environment variable");
    }
    if (!Object.values(endpointResults).every(v => v === true)) {
      console.log("  - Check backend API endpoints in backend/main.py");
    }
  }
  console.log("=".repeat(60) + "\n");

  process.exit(allPassed ? 0 : 1);
}

main();

File: backend/test_api_connection.py
"""
Test API Connection and Endpoints
Tests the FastAPI backend to ensure it's running and endpoints are accessible.
"""

import requests
import sys
from database import engine

# Test Configuration
BACKEND_HOST = "127.0.0.1"
BACKEND_PORT = 8001
BASE_URL = f"http://{BACKEND_HOST}:{BACKEND_PORT}"

def test_database_connection():
    """Test database connectivity"""
    print("\n" + "="*60)
    print("DATABASE CONNECTION TEST")
    print("="*60)
    
    try:
        from sqlalchemy import text
        with engine.connect() as conn:
            result = conn.execute(text("SELECT 1"))
            print("✓ Database connection successful!")
            print(f"✓ Query result: {result.fetchone()}")
            return True
    except Exception as e:
        print(f"✗ Database connection failed: {e}")
        return False


def test_backend_availability():
    """Test if backend API is running"""
    print("\n" + "="*60)
    print("BACKEND AVAILABILITY TEST")
    print("="*60)
    print(f"Testing backend at: {BASE_URL}")
    
    try:
        response = requests.get(f"{BASE_URL}/", timeout=5)
        if response.status_code == 200:
            print("✓ Backend is running!")
            print(f"✓ Response: {response.json()}")
            return True
        else:
            print(f"✗ Backend returned status {response.status_code}")
            return False
    except requests.exceptions.ConnectionError:
        print(f"✗ Cannot connect to backend at {BASE_URL}")
        print("   Make sure the backend is running: python -m uvicorn backend.main:app --reload")
        return False
    except requests.exceptions.Timeout:
        print(f"✗ Backend request timed out")
        return False
    except Exception as e:
        print(f"✗ Backend test failed: {e}")
        return False


def test_api_endpoints():
    """Test key API endpoints"""
    print("\n" + "="*60)
    print("API ENDPOINTS TEST")
    print("="*60)
    
    endpoints = [
        ("GET", "/tools/"),
        ("GET", "/projects/"),
        ("GET", "/interventions/"),
        ("GET", "/tool-types/"),
    ]
    
    results = {}
    for method, endpoint in endpoints:
        try:
            url = f"{BASE_URL}{endpoint}"
            print(f"\nTesting {method} {endpoint}...")
            
            if method == "GET":
                response = requests.get(url, timeout=5)
            
            if response.status_code == 200:
                data = response.json()
                print(f"  ✓ Status 200 OK")
                if isinstance(data, list):
                    print(f"  ✓ Returned {len(data)} items")
                results[endpoint] = True
            else:
                print(f"  ✗ Status {response.status_code}")
                results[endpoint] = False
                
        except requests.exceptions.ConnectionError:
            print(f"  ✗ Cannot connect to {url}")
            results[endpoint] = False
        except Exception as e:
            print(f"  ✗ Error: {e}")
            results[endpoint] = False
    
    return results


def test_cors_headers():
    """Test CORS headers configuration"""
    print("\n" + "="*60)
    print("CORS HEADERS TEST")
    print("="*60)
    
    try:
        response = requests.get(f"{BASE_URL}/", timeout=5)
        headers = response.headers
        
        cors_headers = {
            "access-control-allow-origin": headers.get("access-control-allow-origin", "NOT SET"),
            "access-control-allow-methods": headers.get("access-control-allow-methods", "NOT SET"),
            "access-control-allow-headers": headers.get("access-control-allow-headers", "NOT SET"),
        }
        
        print("\nCORS Headers:")
        for header, value in cors_headers.items():
            print(f"  {header}: {value}")
        
        if headers.get("access-control-allow-origin"):
            print("\n✓ CORS is configured")
            return True
        else:
            print("\n⚠ CORS headers not found (may be configured for preflight)")
            return True
            
    except Exception as e:
        print(f"✗ Error checking CORS: {e}")
        return False


def main():
    """Run all connection tests"""
    print("\n" + "="*60)
    print("APMS CONNECTION TEST SUITE")
    print("="*60)
    
    results = {
        "Database": test_database_connection(),
        "Backend API": test_backend_availability(),
        "CORS": test_cors_headers(),
    }
    
    # Test endpoints only if backend is available
    if results["Backend API"]:
        endpoints = test_api_endpoints()
        results["API Endpoints"] = all(endpoints.values())
    
    # Print summary
    print("\n" + "="*60)
    print("TEST SUMMARY")
    print("="*60)
    
    for test_name, passed in results.items():
        status = "✓ PASS" if passed else "✗ FAIL"
        print(f"{status}: {test_name}")
    
    # Overall result
    all_passed = all(results.values())
    print("\n" + "="*60)
    if all_passed:
        print("✓ ALL TESTS PASSED!")
    else:
        print("✗ SOME TESTS FAILED")
        print("\nTroubleshooting:")
        if not results.get("Database"):
            print("  - Check database URL in .env file")
            print("  - Ensure PostgreSQL is running on 192.168.0.71:5433")
        if not results.get("Backend API"):
            print("  - Start backend: cd APMS && python -m uvicorn backend.main:app --reload")
        if not results.get("CORS"):
            print("  - Check CORS configuration in backend/main.py")
    print("="*60 + "\n")
    
    return 0 if all_passed else 1


if __name__ == "__main__":
    sys.exit(main())
